### DISCUSSION TITLE
Yarn resolutions needed to install @ironclad/rivet-node

### DISCUSSION THREAD

**Discussion by codemile** [Ironclad/rivet#256] [General]
I'm running a large Nx workspace project (monorepo) and had to use the following Yarn resolution to fix a bug, because a dependency was importing `p-queue@7.4.1` instead of `v6`. I'm not sure why this happened, because that dependency had `p-queue@6.6.2`. The offending packages was `rollup-plugin-postcss`.

I traced the conflict back to `@ironclad/rivet-node` and if I uninstalled the package, then everything worked fine. I did some digging and found that it defines these dependencies for `p-queue` so it can install two diff versions.

https://github.com/Ironclad/rivet/blob/b39c02fe8716b7a2147f33083b62988e358f6299/packages/core/package.json#L59

https://github.com/Ironclad/rivet/blob/b39c02fe8716b7a2147f33083b62988e358f6299/packages/core/package.json#L60

I'm starting a discussion, because I'd like to learn more about why this was done? What problem was being solved, etc.

**Comment by abrenneke** [comment]
This was done because p-queue is ESM-only from version 7 up - since `rivet-node` is a dual package that has both bundled ESM and CJS, we couldn't just use p-queue 7 for the CJS bundle. Instead, when bundling the CJS version, we resolve (and inline) p-queue-6 -> p-queue so that we grab the older version for the CJS bundle. The reason that p-queue 6 isn't used everywhere is mainly because the types are broken in p-queue 6.

It's strange that you needed a resolution, because a dependency saying it uses p-queue 7 shouldn't cause other dependencies to start loading it if they only depend on p-queue 6. Is it hoisting p-queue 7 up to the root, and something is doing `require('p-queue')` in a strange place and getting the hoisted version?

**Comment by codemile** [comment]
Yes, p-queue 7 was getting hoisted. I haven't fully tested my project yet to see if my fix is stable, but what you say makes sense to me. I've run into similar pains. My monorepo for example, runs both NextJs and NestJs which have the same issues.


I ended up with this resolution, which seems pretty specific and fixed the build error.

```
  "resolutions": {
    "@nx/rollup/rollup-plugin-postcss/p-queue": "^6.6.2"
  },
  ```
 
I'll leave this here so other people can find it. I don't know why it was needed, but I account it to issues with Nx/Yarn and not this project.

**Reply by abrenneke** [reply]
I feel you. We have over 40 resolutions in our yarn/nx monorepo to fix all sorts of problems
